### ISSUE TITLE
Status of this project

### ISSUE THREAD

**Issue by Yamaha32088** [jaredpalmer/after.js#557]

**Comment by MatthewPattell** [comment]
I do not think that someone will continue to maintain this repository. Although the project is cool in my opinion.
I don't want it to disappear, so I try to update on our fork:
https://github.com/Lomray-Software/after.js
https://www.npmjs.com/package/@lomray/after

react-router v6 + react18.

You need to find the strength in yourself to go to streams 😄

**Comment by Yamaha32088** [comment]
For us just updating the dependencies would be great. We don't necessarily need a lot of features, just bug fixes and keeping up with dependency updates.

**Comment by MatthewPattell** [comment]
@nimaa77 hi,

In my mind I see it like this:
  - Building the project - move away from tsdx to something like rollup, etc. (low priority, but it can be easy)
  - Support for the latest version of react-router and react (medium priority, it can be taken from our fork)
  - The most important part for us is the render to stream (high priority)
  - Update dependencies (medium priority)
  - Fix typescript errors (medium priority, it can be taken from our fork)

Ready to discuss any possibilities and offer any assistance. Therefore, the team and I would very much like the implementation of render to stream.

**Comment by nimaa77** [comment]
I'm on it 😉

**Comment by nimaa77** [comment]
so if we use React 18 it has support for React lazy and code splitting on the server with suspense

also, react-router (remix router) 6.4 has data routers that are SSR ready

This means we can drop support for getInitialProps ([use react-router 6.4 loader feature](https://reactrouter.com/en/main/routers/create-browser-router)) and asyncComponent use React.lazy ([built-in to the react-router](https://reactrouter.com/en/main/route/lazy#lazy))

let me know your thoughts

4 breaking changes in one release 

Drop Support for React 16
Drop Support for React Router 5
Drop Support for asyncComponent (react has native support for this)
Drop Support for getInitialProps (react-router has better, more modernm,

**Comment by nimaa77** [comment]
On the other hand, we can go web server agnostic and use any web server that we want to bootstrap the server (yes this means support for edge runtimes)
having examples that show how to run Vite with after.js can be good as well

**Comment by nimaa77** [comment]
if you think it's a big change, maybe (it might not be possible but I will give it a try)

we can keep all the current APIs and only upgrade react and react-router dependencies, so the list of the breaking changes will become:
1. Drop Support for React 16 - Upgrade to 18
2. Drop Support for React Router 5 - Upgrade to 6.4


then I add new APIs (new functions and components available to import) that include all those nice features that come from react-router 6 (nested layouts, scroll restoration, defer API, loader, and ...)

------------

if it was too much complicated to implement we can break the release into two smaller breaking ones

in the first release, we just upgrade the dependencies
in the second release, we introduce the new APIs and remove the old ones
the best case scenario is to have both of them at the same time to allow the users gradually migrate to new APIs (but this is almost impossible to do)

**Comment by MatthewPattell** [comment]
> so if we use React 18 it has support for React lazy and code splitting on the server with suspense
> 
> react-router (remix router) 6.4 has data routers that are SSR ready This means we can drop support for getInitialProps
> 
> the `transitionBehavior` prop can be removed as well, it was a workaround to control wheater you want to render the next page component while the next page data is loading or if you just want to fetch data at first and then show the next page. react-router 6.4 [Awaited](https://reactrouter.com/en/main/components/await#await) and [Defer](https://reactrouter.com/en/main/utils/defer) APIs will do this (with more control)
> 
> scrollToTop will be dropped as well (since we are going to remove getInitialProps), react-router [Scroll Restoration](https://reactrouter.com/en/main/components/scroll-restoration) will handle that (with way way more control 😍 - this is the best feature)
> 
> so to summerize it would be:
> 
> 6 breaking changes in one release
> 
> 1. Drop Support for React 16 - Upgrade to 18
> 2. Drop Support for React Router 5 - Upgrade to 6.4
> 3. Drop Support for asyncComponent - Alternative: React Lazy, Suspense, [React Router Lazy](https://reactrouter.com/en/main/route/lazy)
> 4. Drop Support for getInitialProps -  Alternative: React Router [Data Apis](https://reactrouter.com/en/main/routers/picking-a-router#data-apis) and Data Router And [Loader](https://reactrouter.com/en/main/route/loader)
> 5. Drop Support for `transitionBehavior` - Alternative [Awaited](https://reactrouter.com/en/main/components/await#await) and [Defer](https://reactrouter.com/en/main/utils/defer)
> 6. 404, After.js will not come with a default 404 anymore
> 
> let me know your thoughts

Well, that sounds even better. In that case, I fully agree with all the points.

I would also like to talk about the SPA and SSR modes. At the moment, After allows for an easy transition from a regular SPA to an SSR, which is often a deciding factor for projects that started without After. We have also made some improvements to After that allow for building the application in both SPA and SSR modes without any changes to functionality.





> if you think it's a big change, maybe (it might not be possible but I will give it a try)
> 
> we can keep all the current APIs and only upgrade react and react-router dependencies, so the list of the breaking changes will become:
> 
> 1. Drop Support for React 16 - Upgrade to 18
> 2. Drop Support for React Router 5 - Upgrade to 6.4
> 
> then I add new APIs (new functions and components available to import) that include all those nice features that come from react-router 6 (nested layouts, scroll restoration, defer API, loader, and ...)
> 
> if it was too much complicated to implement we can break the release into two smaller breaking ones
> 
> in the first release, we just upgrade the dependencies in the second release, we introduce the new APIs and remove the old ones the best case scenario is to have both of them at the same time to allow the users gradually migrate to new APIs (but this is almost impossible to do)

In my opinion, I would focus on the path that allows for the fastest and most efficient transition to new features. In any case, project migration is inevitable, and I would prefer to do it immediately. Ultimately, this should save time.

**Comment by MatthewPattell** [comment]
After giving it a lot of thought, I have come to the conclusion that perhaps we don't need a remix after all. If we want to use it, we can simply go ahead and install it. Additionally, both the remix and the vite integration already exist. 

**I really like this project because it doesn't impose many rules and is not significantly different from a regular React SPA application.** I have studied many solutions for SSR, and all of them immerse us in a sea of rules, thus limiting our freedom of action.

We have tweaked this solution a bit so that switching between spa and ssr happens without modifications. With our current implementation, it became possible to transfer the project, launch, build without pain, fear and anesthesia from SPA to SSR and back.😄

<img width="523" alt="image" src="https://github.com/jaredpalmer/after.js/assets/10459911/60f4d7bd-158f-4ee6-a93a-4401943228c5">

InitialProps - like useEffect for SPA and fetching data for SSR. It works the same in all modes.


Therefore, I believe integrating the latest versions of React and React-DOM with streaming support would be fantastic. Additionally, we could consider using the vite instead of the Razzle but I'm not sure.

I think to breathe new life with the [plan above](https://github.com/jaredpalmer/after.js/issues/557#issuecomment-1491592130)

**Comment by batusai513** [comment]
@MatthewPattell @nimaa77 do you need help with this, I'm actively using this package and would like to get this updates, please let me know so we can move this forward

**Comment by MatthewPattell** [comment]
Right now, I'm torn between two options. On one hand, I want to follow [this plan](https://github.com/jaredpalmer/after.js/issues/557#issuecomment-1517048107), but it would require project migrations, which can be more time-consuming. On the other hand, I could implement [this plan](https://github.com/jaredpalmer/after.js/issues/557#issuecomment-1491592130), which would make the migration of existing projects easier.

Currently, I'm in the process of studying these two plans :) As they say, "the road always appears under the feet of those who walk".

PS: I would also be interested to know if there is any progress from @nimaa77 .

**Comment by MatthewPattell** [comment]
Definitely came to the conclusion that it must be `vite`. The only thing that slows me down is that I can't get `renderToPipeableStream` to work correctly with `createStaticRouter`. I really don't want file navigation. 

There are no problems with `renderToString`.

**Comment by nimaa77** [comment]
@MatthewPattell I had some progress regarding the meta tags and head API


the biggest challenge that I have right now is that I don't know how to map the dynamic modules to the routes so we can send the <script src="path/to/dynamic/route" /> in the initial HTML that we send to the users


Does any one have any ideas about this?

**Comment by MatthewPattell** [comment]
Hi, sorry guys but I got so immersed in solving problems that I came up with my own framework😂. With your permission, I will leave the [link here for study](https://github.com/Lomray-Software/vite-template).

**Comment by golaod** [comment]
Is there a chance to finish up this project to work with router 6 and react 18?
This vite template you gave is awesome but of course it does too much in a same way nextjs does. I don't like having mobx and store managers and all of that. Simple solution like getInitialData was just fine

**Comment by MatthewPattell** [comment]
@golaod The point is, you don't necessarily have to use everything shown there. Just remove MobX, store managers, and anything else you don't need. Then refer to the React Router API v6.

In short: gitInitilProps is replaced with a loader function.
[Here](https://github.com/Lomray-Software/vite-template/blob/prod/src/pages/home/index.tsx#L105) what do you looking for.
